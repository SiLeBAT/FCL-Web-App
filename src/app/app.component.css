:host {
  position: absolute;
  left: 0;
  top: 0;
  width: 100%;
  height: 100%;
  display: flex;
  flex-direction: column;
}

md-sidenav {
  box-shadow: none !important;
  border-left: 1px solid rgba(0, 0, 0, 0.4) !important;
  border-right: 1px solid rgba(0, 0, 0, 0.4) !important;
}

.sidenav-elements {
  padding: 8px 16px;
  display: flex;
  flex-direction: column;
}

.sidenav-elements > * {
  margin: 8px 0;
}

.sidenav-elements > md-select {
  width: 100%;
  margin-top: 20px;
}

.sidenav-elements > button {
  width: 100%;
}

.sidenav-elements > md-radio-group > md-radio-button:not(:last-child) {
  margin-right: 8px;
}

#mainToolbar {
  height: 64px !important;
  justify-content: center !important;
  z-index: 3;
}

#mainContainer {
  width: 100%;
  height: calc(100% - 64px);
}

#mainContent {
  width: 100%;
  height: 100%;
  overflow: hidden;
}

#leftSidenav {
  z-index: 2;
}

#rightSidenav {
  min-width: 200px;
  z-index: 1;
}

#rightSidenavContent {
  width: 100%;
  height: 100%;
  display: flex;
  flex-direction: row;
}

#sidenavSlider {
  width: 10px;
  height: 100%;
  cursor: ew-resize;
  background-color: rgb(200, 200, 200);
  display: flex;
  flex-direction: column;
  justify-content: center;
}

#sidenavSliderCenter {
  width: 100%;
  height: 50px;
  background-color: rgb(100, 100, 100);
}

app-graph {
  width: 100%;
  height: 100%;
}

app-table {
  width: calc(100% - 10px);
  height: 100%;
}
