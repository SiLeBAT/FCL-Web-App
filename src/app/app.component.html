<app-spinner-loader></app-spinner-loader>
<app-alert></app-alert>

<div
  class="master-container"
  fxLayout="column"
  fxLayoutAlign="center stretch">

  <!-- header -->
  <div
    class="header"
    fxFlex>

    <mat-toolbar
      id="mainToolbar"
      class="mat-elevation-z5"
      color="primary">
      <button
        mat-icon-button
        (click)="appService.onToggleLeftSidebar()"
        [disabled]="!isTracingActive()">
        <mat-icon>menu</mat-icon>
      </button>
      <button
        mat-button
        [routerLink]="(isServerLess() || authService.loggedIn()) ? ['/users/main'] : ['/']"
        id="homeButton">
        <span class="mat-title">{{ appName }}</span>
      </button>
      <span
        class="toolbarSpan"
        fxFlex></span>

      <button
        *ngIf="(isServerLess() || authService.loggedIn()) && !isTracingActive()"
        mat-icon-button
        [routerLink]="['/tracing']"
        matTooltip="Tracing view">
        <mat-icon>timeline</mat-icon>

      </button>

      <button
        *ngIf="(isServerLess() || authService.loggedIn()) && isTracingActive()"
        mat-icon-button
        (click)="appService.onSaveImage()"
        matTooltip="Save image">
        <mat-icon>image</mat-icon>
      </button>
      <input
        *ngIf="(isServerLess() || authService.loggedIn()) && isTracingActive()"
        id="fileInput"
        #fileInput
        type="file"
        (change)="appService.onLoad($event)">
      <button
        *ngIf="(isServerLess() || authService.loggedIn()) && isTracingActive()"
        mat-icon-button
        onclick="document.getElementById('fileInput').click()"
        matTooltip="Upload data">
        <mat-icon>file_upload</mat-icon>
      </button>
      <button
        *ngIf="(isServerLess() || authService.loggedIn()) && isTracingActive()"
        mat-icon-button
        (click)="appService.onSave()"
        matTooltip="Download data">
        <mat-icon>file_download</mat-icon>
      </button>
      <button
        *ngIf="(isServerLess() || authService.loggedIn()) && isTracingActive()"
        mat-icon-button
        (click)="appService.onVisioLayout()"
        matTooltip="Generate Visio Layout">
        <mat-icon>image</mat-icon>
      </button>
      <button
        *ngIf="!rightSidenav.opened && (isServerLess() || authService.loggedIn()) && isTracingActive()"
        mat-icon-button
        (click)="appService.onToggleRightSidebar()"
        matTooltip="Toggle right sidebar">
        <mat-icon>info</mat-icon>
      </button>
      <button
        *ngIf="rightSidenav.opened && (isServerLess() || authService.loggedIn()) && isTracingActive()"
        mat-mini-fab
        (click)="appService.onToggleRightSidebar()"
        matTooltip="Toggle right sidebar">
        <mat-icon>info</mat-icon>
      </button>
      <mat-divider [inset]="true" [vertical]="true"></mat-divider>

      <a
        *ngIf="!authService.loggedIn()"
        mat-button
        [routerLink]="['/users/login']">
        <mat-icon>exit_to_app</mat-icon>
        Login/Register
      </a>

      <button
        *ngIf="authService.loggedIn()"
        mat-icon-button
        [matMenuTriggerFor]="menu">
        <mat-icon>account_circle</mat-icon>
      </button>
      <mat-menu
        #menu="matMenu"
        [overlapTrigger]="false">
        <div mat-menu-item disabled>
          <mat-icon>account_circle</mat-icon>
          <span>{{ getCurrentUserEmail() }}</span>
        </div>
        <mat-divider></mat-divider>
        <button
          mat-menu-item
          (click)="logout()">
          <mat-icon>power_settings_new</mat-icon>
          <span>Logout</span>
        </button>
      </mat-menu>


    </mat-toolbar>
  </div>


  <!-- main content -->
  <div
    class="main-content"
    fxFlex="90vh">

    <mat-sidenav-container
      id="sidenavContainer"
      #sidenavContainer>

      <!-- left sidenav -->
      <mat-sidenav
        id="leftSidenav"
        #leftSidenav
        position="start"
        mode="side">
        <h1>mat-sidenav</h1>
      </mat-sidenav>

      <!-- main content -->
      <mat-sidenav-content>
        <router-outlet></router-outlet>
      </mat-sidenav-content>

      <!-- right sidenav -->
      <mat-sidenav
        id="rightSidenav"
        #rightSidenav
        position="end"
        mode="side">
        <div id="rightSidenavContent">
          <div id="sidenavSlider" #sidenavSlider>
            <div id="sidenavSliderCenter"></div>
          </div>
          <app-table #table></app-table>
        </div>
      </mat-sidenav>

    </mat-sidenav-container>
  </div>


  <!-- footer -->
  <div
    class="footer"
    fxFlex>
    <mat-toolbar
      class="mat-elevation-z5"
      color="accent"
      fxLayout="row"
      fxLayoutAlign="center center">
      <div class="footer-element" fxFlex>
        <a
          mat-button
          href="https://www.bfr.bund.de"
          target="_blank">
          <mat-icon>public</mat-icon>
          <span class="mat-caption"> BfR - Bundesinstitut für Risikobewertung</span>
          <!-- BfR - Bundesinstitut für Risikobewertung -->
        </a>
      </div>
      <div class="footer-element" fxFlex>
        <a
          mat-button
          href="https://foodrisklabs.bfr.bund.de"
          target="_blank">
          <mat-icon>public</mat-icon>
          <span class="mat-caption"> FoodRisk - Labs</span>
        </a>
      </div>
      <div class="footer-element" fxFlex>
        <a
          mat-button
          href="mailto:{{ supportContact }}"
          target="_top">
          <mat-icon>mail</mat-icon>
          <span class="mat-caption"> Report a Problem</span>
        </a>
      </div>
      <a [routerLink]="['/graph-editor']" style="display:block; height:25px; width:25px"> </a>
    </mat-toolbar>
  </div>

</div>
