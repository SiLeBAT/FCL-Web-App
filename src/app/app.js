'use strict';

/*global angular*/

angular.module('app', ['ngMaterial', 'ui.router', 'ngResource', 'md.data.table'])
    .config(function($mdIconProvider, $mdThemingProvider) {
        var primary = 'indigo';
        var accent = 'pink';
        var warn = 'red';
        var background = 'grey';

        $mdThemingProvider.theme('default').primaryPalette(primary).accentPalette(accent).warnPalette(warn).backgroundPalette(background);
        $mdThemingProvider.theme('dark').primaryPalette(primary).accentPalette(accent).warnPalette(warn).backgroundPalette(background).dark();
        $mdIconProvider.icon('close', './icons/ic_close_white_24px.svg', 24);
    }).config(function($stateProvider, $urlRouterProvider) {
        $stateProvider
            .state('graph', {
                url: '/',
                template: '<graph-view></graph-view>'
            })
            .state('table', {
                url: '/table',
                template: '<table-view></table-view>'
            })
            .state('split', {
                url: '/split',
                template: '<split-view></split-view>'
            });

        $urlRouterProvider.otherwise('/');
    });
