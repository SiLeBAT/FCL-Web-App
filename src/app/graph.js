'use strict';

/*global angular, cytoscape, $*/

angular.module('app').service('graph', function(graphComputations, $mdDialog) {

  var graph = this;

  var cy;

  var fontSize;

  graph.init = function(data) {
    cy = cytoscape({
      container: $('#graph')[0],

      style: cytoscape.stylesheet()
        .selector('node')
        .css({
          'content': 'data(name)',
          'height': '50',
          'width': '50',
          'background-color': '#FFFFFF',
          'border-width': 3,
          'border-color': '#000000',
          'text-valign': 'bottom',
          'text-halign': 'right',
          'color': '#000000'
        })
        .selector('edge')
        .css({
          'target-arrow-shape': 'triangle',
          'width': 6,
          'line-color': '#000000',
          'target-arrow-color': '#FF0000',
          'curve-style': 'bezier'
        })
        .selector('node:selected')
        .css({
          'background-color': '#8080FF',
          'border-width': 6,
          'border-color': '#0000FF',
          'color': '#0000FF'
        })
        .selector('edge:selected')
        .css({
          'width': 12,
          'line-color': '#00FF00',
          'target-arrow-color': '#FF0000'
        }).selector('node[?forward], node[?backward]')
        .css({
          'background-color': '#00FF00'
        }).selector('node:selected[?forward], node:selected[?backward]')
        .css({
          'background-color': '#008080'
        }).selector('edge[?forward], edge[?backward]')
        .css({
          'line-color': '#00FF00'
        }),

      layout: {
        name: 'cose'
      },

      elements: {
        nodes: data.stations,
        edges: data.deliveries
      },

      zoom: 1,
      pan: {
        x: 0,
        y: 0
      },

      minZoom: 0.3,
      maxZoom: 3,

      wheelSensitivity: 0.5,

      ready: function() {
        graph.setFontSize(12);

        var relations = data.deliveriesRelations;

        cy.batch(function() {
          cy.edges().data('from', []);
          cy.edges().data('to', []);

          for (var i = 0, j = relations.length; i < j; i++) {
            var from = relations[i].data.from;
            var to = relations[i].data.to;
            var fromDelivery = cy.$('#' + from);
            var toDelivery = cy.$('#' + to);

            fromDelivery.data('to', fromDelivery.data('to').concat(to));
            toDelivery.data('from', toDelivery.data('from').concat(from));
          }
        });
      }
    });

    cy.on('zoom', function(event) {
      graph.setFontSize(fontSize);
    });

    cy.cxtmenu({
      selector: 'node',
      commands: [{
        content: 'Show Forward Trace',
        select: function(station) {
          cy.batch(function() {
            graphComputations.clearForwardTrace();
            graphComputations.clearBackwardTrace();
            graphComputations.showStationForwardTrace(station);
          });
        }
      }, {
        content: 'Show Backward Trace',
        select: function(station) {
          cy.batch(function() {
            graphComputations.clearForwardTrace();
            graphComputations.clearBackwardTrace();
            graphComputations.showStationBackwardTrace(station);
          });
        }
      }]
    });

    cy.cxtmenu({
      selector: 'edge',
      commands: [{
        content: 'Show Forward Trace',
        select: function(delivery) {
          cy.batch(function() {
            graphComputations.clearForwardTrace();
            graphComputations.clearBackwardTrace();
            graphComputations.showDeliveryForwardTrace(delivery);
          });
        }
      }, {
        content: 'Show Backward Trace',
        select: function(delivery) {
          cy.batch(function() {
            graphComputations.clearForwardTrace();
            graphComputations.clearBackwardTrace();
            graphComputations.showDeliveryBackwardTrace(delivery);
          });
        }
      }]
    });

    cy.cxtmenu({
      selector: 'core',
      commands: [{
        content: 'bg1',
        select: function() {
          $mdDialog.show({
            controller: function($scope, $mdDialog, data) {
              $scope.data = data;

              $scope.closeDialog = function() {
                $mdDialog.hide();
              };
            },
            templateUrl: 'app/dialog.template.html',
            parent: angular.element(document.body),
            clickOutsideToClose: true,
            locals: {
              data: data
            }
          });
        }
      }, {
        content: 'bg2',
        select: function() {}
      }]
    });

    graphComputations.init(cy);
  };

  graph.setNodeSize = function(size) {
    cy.nodes().css({
      'height': size,
      'width': size
    });
  };

  graph.setFontSize = function(size) {
    fontSize = size;

    cy.nodes().css({
      'font-size': Math.max(fontSize / cy.zoom(), fontSize)
    });
  };

});