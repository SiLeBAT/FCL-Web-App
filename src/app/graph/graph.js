'use strict';

/*global angular, cytoscape, $*/

angular.module('app').service('graph', function(tracing, $mdDialog) {

  var graph = this;

  var cy;
  var fontSize;

  graph.initFromData = function(data) {
    cy = cytoscape({
      container: $('#graph')[0],

      elements: {
        nodes: data.stations,
        edges: data.deliveries
      },

      layout: {
        name: 'cose-bilkent'
      },

      style: style,
      minZoom: 0.1,
      maxZoom: 10,
      wheelSensitivity: 0.5,
    });

    init();
  };

  graph.initFromJson = function(json) {
    cy = cytoscape({
      container: $('#graph')[0],

      elements: json.elements,

      layout: {
        name: 'preset',
        zoom: json.zoom,
        pan: json.pan
      },

      style: style,
      minZoom: json.minZoom,
      maxZoom: json.maxZoom,
      wheelSensitivity: json.wheelSensitivity,
    });

    init();
  };

  graph.getJson = function() {
    if (cy !== undefined) {
      return cy.json();
    }
    else {
      return undefined;
    }
  };

  graph.setNodeSize = function(size) {
    cy.nodes().css({
      'height': size,
      'width': size
    });
  };

  graph.setFontSize = function(size) {
    fontSize = size;

    cy.nodes().css({
      'font-size': Math.max(fontSize / cy.zoom(), fontSize)
    });
  };

  var init = function() {
    cy.on('zoom', function(event) {
      graph.setFontSize(fontSize);
    });
    cy.cxtmenu(contextMenu);
    cy.cxtmenu(stationContextMenu);
    cy.cxtmenu(deliveryContextMenu);
    tracing.init(cy);
  };

  var style = cytoscape.stylesheet()
    .selector('node')
    .css({
      'content': 'data(name)',
      'background-color': '#FFFFFF',
      'border-width': 3,
      'border-color': '#000000',
      'text-valign': 'bottom',
      'text-halign': 'right',
      'color': '#000000'
    })
    .selector('edge')
    .css({
      'target-arrow-shape': 'triangle',
      'width': 6,
      'line-color': '#000000',
      'target-arrow-color': '#FF0000',
      'curve-style': 'bezier'
    })
    .selector('node:selected')
    .css({
      'background-color': '#8080FF',
      'border-width': 6,
      'border-color': '#0000FF',
      'color': '#0000FF'
    })
    .selector('edge:selected')
    .css({
      'width': 12,
      'line-color': '#00FF00',
      'target-arrow-color': '#FF0000'
    }).selector('node[?forward], node[?backward]')
    .css({
      'background-color': '#00FF00'
    }).selector('node:selected[?forward], node:selected[?backward]')
    .css({
      'background-color': '#008080'
    }).selector('edge[?forward], edge[?backward]')
    .css({
      'line-color': '#00FF00'
    });

  var contextMenu = {
    selector: 'core',
    commands: [{
      content: 'Apply Layout',
      select: function() {
        $mdDialog.show({
          controller: function($scope) {
            $scope.select = function(layout) {
              $mdDialog.hide(layout);
            };
          },
          template: '<md-dialog aria-label="Apply Layout">' +
            '<md-toolbar><dialog-toolbar title="Apply Layout"></dialog-toolbar></md-toolbar>' +
            '<md-dialog-content><layout-select on-select="select(layout)"></layout-select></md-dialog-content>' +
            '</md-dialog>',
          parent: angular.element(document.body),
          clickOutsideToClose: true
        }).then(function(layout) {
          cy.layout({
            name: layout,
            animate: true
          });
        });
      }
    }, {
      content: 'Zoom to Graph',
      select: function() {
        cy.fit();
      }
    }]
  };

  var stationContextMenu = {
    selector: 'node',
    commands: [{
      content: 'Show Forward Trace',
      select: function(station) {
        cy.batch(function() {
          tracing.clearForwardTrace();
          tracing.clearBackwardTrace();
          tracing.showStationForwardTrace(station);
        });
      }
    }, {
      content: 'Show Backward Trace',
      select: function(station) {
        cy.batch(function() {
          tracing.clearForwardTrace();
          tracing.clearBackwardTrace();
          tracing.showStationBackwardTrace(station);
        });
      }
    }]
  };

  var deliveryContextMenu = {
    selector: 'edge',
    commands: [{
      content: 'Show Forward Trace',
      select: function(delivery) {
        cy.batch(function() {
          tracing.clearForwardTrace();
          tracing.clearBackwardTrace();
          tracing.showDeliveryForwardTrace(delivery);
        });
      }
    }, {
      content: 'Show Backward Trace',
      select: function(delivery) {
        cy.batch(function() {
          tracing.clearForwardTrace();
          tracing.clearBackwardTrace();
          tracing.showDeliveryBackwardTrace(delivery);
        });
      }
    }]
  };
});
