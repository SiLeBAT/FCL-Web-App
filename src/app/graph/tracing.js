'use strict';

/*global angular*/

angular.module('app').service('tracing', function() {

    var comp = this;

    var cy;

    comp.init = function(cytoscape) {
        cy = cytoscape;
    };

    comp.clearForwardTrace = function() {
        cy.$().data('forward', false);
    };

    comp.clearBackwardTrace = function() {
        cy.$().data('backward', false);
    };

    comp.showStationForwardTrace = function(station) {
        station.data('forward', true);
        cy.$('edge[source = "' + station.id() + '"]').forEach(function(d) {
            comp.showDeliveryForwardTrace(d);
        });
    };

    comp.showStationBackwardTrace = function(station) {
        station.data('backward', true);
        cy.$('edge[target = "' + station.id() + '"]').forEach(function(d) {
            comp.showDeliveryBackwardTrace(d);
        });
    };
    
    comp.showDeliveryForwardTrace = function(delivery) {
        delivery.data('forward', true);
        cy.$('#' + delivery.data('target')).data('forward', true);

        var to = delivery.data('to');

        cy.filter(function(i, e) {
            return e.isEdge() && to.includes(e.id());
        }).forEach(function(d) {
            comp.showDeliveryForwardTrace(d);
        });
    };

    comp.showDeliveryBackwardTrace = function(delivery) {
        delivery.data('backward', true);
        cy.$('#' + delivery.data('source')).data('backward', true);

        var from = delivery.data('from');

        cy.filter(function(i, e) {
            return e.isEdge() && from.includes(e.id());
        }).forEach(function(d) {
            comp.showDeliveryBackwardTrace(d);
        });
    };
});
