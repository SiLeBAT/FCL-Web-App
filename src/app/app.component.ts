import {Component, OnInit, ViewChild} from '@angular/core';
import {MdDialog, MdSidenav} from '@angular/material';

import {GraphComponent} from './graph/graph.component';
import {TableComponent} from './table/table.component';
import {DataService} from './util/data.service';
import {DialogSelectComponent, DialogSelectData} from './dialog/dialog-select/dialog-select.component';
import {Utils} from './util/utils';
import {FclData, FclElements, GraphType, TableMode} from './util/datatypes';
import {Constants} from './util/constants';
import {GisComponent} from './gis/gis.component';

@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.css']
})
export class AppComponent implements OnInit {

  @ViewChild('graph') graph: GraphComponent;
  @ViewChild('gis') gis: GisComponent;
  @ViewChild('table') table: TableComponent;
  @ViewChild('rightSidenav') rightSidenav: MdSidenav;

  graphTypes = Constants.GRAPH_TYPES;
  graphType = GraphType.GRAPH;
  gisType = GraphType.GIS;
  tableModes = Constants.TABLE_MODES;
  showTypes = Constants.SHOW_TYPES;
  sizes = Constants.SIZES;

  data: FclData = {
    elements: null,
    layout: null,
    gisLayout: null,
    graphSettings: DataService.getDefaultGraphSettings(),
    tableSettings: DataService.getDefaultTableSettings()
  };

  constructor(private dataService: DataService, private dialogService: MdDialog) {
    document.body.oncontextmenu = e => e.preventDefault();
  }

  ngOnInit() {
    this.dataService.setDataSource('assets/data/bbk.json');
    this.dataService.getData().then(data => {
      this.data = data;
      this.updateComponents();
    }).catch(error => {
      Utils.showErrorMessage(this.dialogService, error);
    });

    this.rightSidenav.onOpenStart.subscribe(() => this.onTableChange('width'));
    new Hammer(document.getElementById('sidenavSlider')).on('pan', event => {
      const newWidth = 1 - event.center.x / document.getElementById('mainContainer').offsetWidth;

      if (newWidth > 0 && newWidth < 1) {
        this.data.tableSettings.width = newWidth;
        this.onTableChange('width');
      }
    });
  }

  onGraphChange(property: string) {
    switch (property) {
      case 'type':
        switch (this.data.graphSettings.type) {
          case GraphType.GRAPH:
            this.data.gisLayout = this.gis.getLayout();
            break;
          case GraphType.GIS:
            this.data.layout = this.graph.getLayout();
            break;
        }

        this.updateComponents();
        break;
      case 'nodeSize':
        this.getCurrentGraph().setNodeSize(this.data.graphSettings.nodeSize);
        break;
      case 'fontSize':
        this.getCurrentGraph().setFontSize(this.data.graphSettings.fontSize);
        break;
      case 'mergeDeliveries':
        this.getCurrentGraph().setMergeDeliveries(this.data.graphSettings.mergeDeliveries);
        break;
      case 'showLegend':
        this.getCurrentGraph().setShowLegend(this.data.graphSettings.showLegend);
        break;
      case 'showZoom':
        this.getCurrentGraph().setShowZoom(this.data.graphSettings.showZoom);
        break;
    }
  }

  onTableChange(property: string) {
    switch (property) {
      case 'width':
        document.getElementById('rightSidenav').style.width = (this.data.tableSettings.width * 100) + '%';
        break;
      case 'mode':
        this.table.setMode(this.data.tableSettings.mode);
        break;
      case 'stationColumns':
        this.table.setStationColumns(this.data.tableSettings.stationColumns);
        break;
      case 'deliveryColumns':
        this.table.setDeliveryColumns(this.data.tableSettings.deliveryColumns);
        break;
      case 'showType':
        this.table.setShowType(this.data.tableSettings.showType);
        break;
    }
  }

  onLoad(event) {
    const files: FileList = event.target.files;

    if (files.length === 1) {
      this.dataService.setDataSource(files[0]);
      this.dataService.getData().then(data => {
        this.data = data;
        this.updateComponents();
      }).catch(error => {
        Utils.showErrorMessage(this.dialogService, error);
      });
    } else {
      Utils.showErrorMessage(this.dialogService, 'Please select one .json file!');
    }

    (<HTMLInputElement>document.getElementById('fileInput')).value = '';
  }

  onSave() {
    switch (this.data.graphSettings.type) {
      case GraphType.GRAPH:
        this.data.layout = this.graph.getLayout();
        break;
      case GraphType.GIS:
        this.data.gisLayout = this.gis.getLayout();
        break;
    }

    const blob = new Blob([JSON.stringify(this.data)], {type: 'application/json'});
    const fileName = 'data.json';

    if (window.navigator.msSaveOrOpenBlob != null) {
      window.navigator.msSaveOrOpenBlob(blob, fileName);
    } else {
      const url = window.URL.createObjectURL(blob);

      Utils.openSaveDialog(url, fileName);
      window.URL.revokeObjectURL(url);
    }
  }

  onSaveImage() {
    this.getCurrentGraph().getCanvas().then(canvas => {
      const fileName = 'graph.png';

      if (window.navigator.msSaveOrOpenBlob != null && canvas.msToBlob != null) {
        window.navigator.msSaveOrOpenBlob(canvas.msToBlob(), fileName);
      } else {
        Utils.openSaveDialog(canvas.toDataURL('image/png'), fileName);
      }
    });
  }

  changeColumns() {
    const options: { value: string, viewValue: string, selected: boolean }[] = [];

    for (const column of Utils.getAllTableProperties(this.data.tableSettings.mode, this.data.elements)) {
      options.push({
        value: column,
        viewValue: Constants.PROPERTIES.has(column) ? Constants.PROPERTIES.get(column).name : '"' + column + '"',
        selected: Utils.getTableProperties(this.data.tableSettings.mode, this.data.tableSettings.stationColumns,
          this.data.tableSettings.deliveryColumns).includes(column)
      });
    }

    const dialogData: DialogSelectData = {
      title: 'Input',
      options: options
    };

    this.dialogService.open(DialogSelectComponent, {data: dialogData}).afterClosed().subscribe(selections => {
      if (selections != null) {
        switch (this.data.tableSettings.mode) {
          case TableMode.STATIONS:
            this.data.tableSettings.stationColumns = selections;
            this.onTableChange('stationColumns');
            break;
          case TableMode.DELIVERIES:
            this.data.tableSettings.deliveryColumns = selections;
            this.onTableChange('deliveryColumns');
            break;
        }
      }
    });
  }

  private updateComponents() {
    document.getElementById('rightSidenav').style.width = (this.data.tableSettings.width * 100) + '%';

    switch (this.data.graphSettings.type) {
      case GraphType.GRAPH:
        const waitForGraph = () => {
          if ((<HTMLElement>document.getElementsByTagName('app-graph')[0]).offsetParent == null) {
            setTimeout(waitForGraph, 50);
          } else {
            this.updateGraphAndTable();
          }
        };

        waitForGraph();
        break;
      case GraphType.GIS:
        const waitForGis = () => {
          if ((<HTMLElement>document.getElementsByTagName('app-gis')[0]).offsetParent == null) {
            setTimeout(waitForGis, 50);
          } else {
            this.updateGraphAndTable();
          }
        };

        waitForGis();
        break;
    }
  }

  private updateGraphAndTable() {
    this.getCurrentGraph().setNodeSize(this.data.graphSettings.nodeSize);
    this.getCurrentGraph().setFontSize(this.data.graphSettings.fontSize);
    this.getCurrentGraph().setMergeDeliveries(this.data.graphSettings.mergeDeliveries);
    this.getCurrentGraph().setShowLegend(this.data.graphSettings.showLegend);
    this.getCurrentGraph().onChange(() => this.table.update());

    switch (this.data.graphSettings.type) {
      case GraphType.GRAPH:
        this.graph.init(this.data.elements, this.data.layout);
        break;
      case GraphType.GIS:
        this.gis.init(this.data.elements, this.data.gisLayout);
        break;
    }

    this.table.setMode(this.data.tableSettings.mode);
    this.table.setStationColumns(this.data.tableSettings.stationColumns);
    this.table.setDeliveryColumns(this.data.tableSettings.deliveryColumns);
    this.table.setShowType(this.data.tableSettings.showType);
    this.table.onSelectionChange(() => this.getCurrentGraph().updateSelection());
    this.table.init(this.data.elements);
  }

  private getCurrentGraph(): GraphComponent | GisComponent {
    switch (this.data.graphSettings.type) {
      case GraphType.GRAPH:
        return this.graph;
      case GraphType.GIS:
        return this.gis;
    }
  }
}
