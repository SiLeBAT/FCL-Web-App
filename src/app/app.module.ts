import { BrowserModule } from '@angular/platform-browser';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
import { AppRoutingModule } from './app.routing.module';
import { NgModule } from '@angular/core';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { FlexLayoutModule } from '@angular/flex-layout';
import { LocationStrategy, HashLocationStrategy } from '@angular/common';
import { StoreModule } from '@ngrx/store';
import { StoreDevtoolsModule } from '@ngrx/store-devtools';
import { EffectsModule } from '@ngrx/effects';

import { NgxDatatableModule } from '@swimlane/ngx-datatable';
import { ScrollbarHelper } from '@swimlane/ngx-datatable/release/services/scrollbar-helper.service';
import { PasswordStrengthMeterModule } from 'angular-password-strength-meter';

import { UserModule } from './user/user.module';
import { CoreModule } from './core/core.module';
import { SharedModule } from './shared/shared.module';
import { MaterialModule } from './shared/material.module';
import { GraphEditorModule } from './graph-editor/graph-editor.module';

import { AppComponent } from './app.component';
import { TracingComponent } from './tracing/tracing.component';
import { GraphComponent } from './graph/graph.component';
import { GisComponent } from './gis/gis.component';
import { TableComponent } from './table/table.component';
import { DialogActionsComponent } from './dialog/dialog-actions/dialog-actions.component';
import { DialogAlertComponent } from './dialog/dialog-alert/dialog-alert.component';
import { DialogPromptComponent } from './dialog/dialog-prompt/dialog-prompt.component';
import { DialogSelectComponent } from './dialog/dialog-select/dialog-select.component';
import { DialogSingleSelectComponent } from './dialog/dialog-single-select/dialog-single-select.component';
import { StationPropertiesComponent } from './dialog/station-properties/station-properties.component';
import { DeliveryPropertiesComponent } from './dialog/delivery-properties/delivery-properties.component';
import { VisioLayoutComponent } from './visio/visio-dialog/visio-dialog.component';
import { environment } from '../environments/environment';
import { STATE_SLICE_NAME, reducer } from './state/tracing.reducers';
import { TracingEffects } from './state/tracing.effects';

@NgModule({
    declarations: [
        AppComponent,
        TracingComponent,
        GraphComponent,
        GisComponent,
        TableComponent,
        DialogActionsComponent,
        DialogAlertComponent,
        DialogPromptComponent,
        DialogSelectComponent,
        DialogSingleSelectComponent,
        StationPropertiesComponent,
        DeliveryPropertiesComponent,
        VisioLayoutComponent
    ],
    imports: [
        BrowserModule,
        BrowserAnimationsModule,
        FormsModule,
        ReactiveFormsModule,
        MaterialModule,
        NgxDatatableModule,
        GraphEditorModule,
        FlexLayoutModule,
        PasswordStrengthMeterModule,
        UserModule,
        CoreModule,
        SharedModule,
        MaterialModule,
        StoreModule.forRoot({ [STATE_SLICE_NAME]: reducer }),
        EffectsModule.forRoot([TracingEffects]),
        StoreDevtoolsModule.instrument({
            name: 'FCL Devtools',
            maxAge: 25,
            logOnly: environment.production
        }),
        // AppRoutingModule needs to be at the end
        AppRoutingModule
    ],
    providers: [
        // {
        //     provide: LocationStrategy,
        //     useClass: HashLocationStrategy
        // },
        ScrollbarHelper
    ],
    bootstrap: [
        AppComponent
    ],
    entryComponents: [
        DialogActionsComponent,
        DialogAlertComponent,
        DialogPromptComponent,
        DialogSelectComponent,
        DialogSingleSelectComponent,
        StationPropertiesComponent,
        DeliveryPropertiesComponent,
        VisioLayoutComponent
    ]
})
export class AppModule {
}
