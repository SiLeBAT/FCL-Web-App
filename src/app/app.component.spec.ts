import {async, ComponentFixture, TestBed} from '@angular/core/testing';
import {BrowserAnimationsModule} from '@angular/platform-browser/animations';
import {MaterialModule} from '@angular/material';
import {FormsModule} from '@angular/forms';
import {NgxDatatableModule} from '@swimlane/ngx-datatable';
import {GraphComponent} from './graph/graph.component';
import {TableComponent} from './table/table.component';
import {DataService} from './util/data.service';
import {TracingService} from './graph/tracing.service';

import {AppComponent} from './app.component';

describe('AppComponent', () => {
  let component: AppComponent;
  let fixture: ComponentFixture<AppComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [
        BrowserAnimationsModule,
        MaterialModule,
        FormsModule,
        NgxDatatableModule
      ],
      declarations: [
        AppComponent,
        GraphComponent,
        TableComponent
      ],
      providers: [
        DataService,
        TracingService
      ]
    }).compileComponents().then();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(AppComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should render title in a md-toolbar', async(() => {
    const compiled = fixture.debugElement.nativeElement;
    expect(compiled.querySelector('md-toolbar').textContent).toContain('FoodChain-Lab');
  }));
});
