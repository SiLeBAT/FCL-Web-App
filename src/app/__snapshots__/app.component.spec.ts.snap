// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppComponent should render correctlu 1`] = `
<fcl-root>
  <fcl-main-page />
</fcl-root>
`;
